### PR TITLE
Feat/http client events

### PR DESCRIPTION
**Pull request by ADmad**
Refs #17005, alternative to #17366

### CODE REVIEW

**Comment by LordSimal** [comment]
maybe I am too dumb to understand but besides setting the eventClass property and reusing the data property for holding the data what exactly is the benefit of doing it this way?

**Comment by ADmad** [comment]
> maybe I am too dumb to understand but besides setting the eventClass property and reusing the data property for holding the data what exactly is the benefit of doing it this way?

If you ignore the implementation details not much :)

- Using `EventDispatcherTrait::dispatchEvent()` instead of creating the event instance yourself maintains code consistency across the framework. `EventDispatcherTrait::_eventClass` exits for the purpose of being able to switch the event class.
- While I can't think of any use case, if someone did `$client->getEventManager()->dispatchEvent()` it would create an event with the wrong class. Which makes think that `dispatchEvent()` should be made protected.
- In your PR you call the setter methods of the event class right after object creating. In general such usage is an indicator that the required context should be passed through the constructor itself.
- The other PR doesn't for response instance modified in `HttpClient.afterSend`. asgraf mentioned the potential use case "replacing response object here may have some niche use case like extra response content validation".

I have spend way more time commenting on these PRs than writing the actual code changes :)

**Comment by LordSimal** [comment]
I appreciate the time you spend explaining this. But here are some of my thoughts why I chose my implementation

---

> Using EventDispatcherTrait::dispatch() instead of creating the event instance yourself maintains code consistency across the framework. EventDispatcherTrait::_eventClass exits for the purpose of being able to switch the event class.

I first tried to overwrite the property inside the Client class, but this is not allowed by PHP. Setting the property manually in the constructor was not something I wanted to do because having to extend the Client's constructor just to change the Event class also felt weird (I personally would add a `setEventClass()` method in the trait but there surely is another reason why this is not the case)

--- 

> In your PR you call the setter methods of the event class right after object creating. In general such usage is an indicator that the required context should be passed through the constructor itself.

I wanted to use the same methods as the user would use when listening to the events, so there are no 2 different ways of setting data in the event.

---

> The other PR doesn't for response instance modified in HttpClient.afterSend.

yup, I missed that one because I expected replacing the response in the `beforeSend` was enough.

---

> I have spend way more time commenting on these PRs than writing the actual code changes :)

Yes, that may be the case, but at least I am trying to learn and understand why this design is chosen/better compared to what I have tried before. I still have to learn much, sure, but as you can see this kind of stuff is not as 100% clear and self-explanatory as it may be to you. Especially since these kinds of core internal design choices are not really documented anywhere.

I wouldn't have minded if you just added some commits into the already existing branch and adjusted it accordingly, but I guess it was easier for you to just create a new one.

**Comment by othercorey** [comment]
@LordSimal I think the benefit of a separate PR was to show the changes without forcibly modifying your versions so this discussion can happen.  I don't think we'd invalidate your concerns just because of the new proposal sunless there's a simple consensus. As you noted, getting this kind of API and implementation right can be tricky and there is historical context.

**Comment by LordSimal** [comment]
In the end if @asgraf can achieve his goals with this PR and @markstory is also fine with these changes I am also happy.

The only thing I have learned is that its best if my participation in new features keeps to commenting on the issue since my implementation PR's always have 30+ comments + several force pushes till they are at a somewhat correct state of where it needs to be for this framework. Don't want to waste anyones time as can be seen by ADmads comment above.

**Comment by othercorey** [comment]
It's just what happens when we open PRs to start a discussion. Either there is a lengthy discussion in the PR or we have to re-write the feature after committing. This happened with several other recent features.

File: src/Http/Client.php [hunk 1]
```diff
@@ -479,7 +489,25 @@ public function send(RequestInterface $request, array $options = []): Response
...
+            if ($response === null) {
+                $response = $this->_sendRequest($request, $event->getOptions());
+
+                /** @var \Cake\Http\Client\ClientEvent $event */
+                $event = $this->dispatchEvent(
+                    'HttpClient.afterSend',
+                    ['request' => $request, 'options' => $options, 'response' => $response]
+                );
+                $response = $event->getResult();
+                assert($response instanceof Response);
```
Comment (markstory) [inline src/Http/Client.php]: Part of me wants us to always call the `afterSend` event. Without it always being called it makes the events hard to use for telemetry purposes, which I think could be a common use case for these events, alongside caching and logging.

File: tests/TestCase/Http/ClientTest.php [hunk 2]
```diff
@@ -841,6 +842,72 @@ public function testSendRequest(): void
...
     }
 
+    public function testBeforeSend(): void
+    {
+        $client = new Client();
+        $client->getEventManager()->on(
+            'HttpClient.beforeSend',
+            function (ClientEvent $event, Request $request, array $options) {
+                $this->assertSame('http://foo.test', (string)$request->getUri());
+                $this->assertSame('thing', $options['some']);
```
Comment (markstory) [inline tests/TestCase/Http/ClientTest.php]: This test would pass even if this callback was never triggered. We could use a local variable that is mutated by the callback to ensure it was called.

Something like

```php
        $callbackCalled = false;
        $client->getEventManager()->on(
            'HttpClient.beforeSend',
            function (ClientEvent $event, Request $request, array $options) use (&$callbackCalled) {
                $callbackCalled = true;
                // ...
             }
         );
        $this->assertTrue($callbackCalled);
````

**Comment by othercorey** [comment]
@markstory I think new person evaluating would help here.

**Comment by markstory** [comment]
> (I personally would add a setEventClass() method in the trait but there surely is another reason why this is not the case)

I don't think any had proposed it yet. We often try to keep public APIs as small as possible. It keeps our footprint smaller and easier to maintain. We can always expand those interfaces in the future as the need arises.

> Don't want to waste anyones time as can be seen by ADmads comment above.

I don't think your approach was a waste of anyone's time. While it may not be the solution that is taken, it helped get us there. Sometimes the alternate path isn't apparent until you see another solution. Unfortunately our process isn't as direct as it can be in coworker environment as our communication has much higher latency.

File: src/Http/Client/ClientEvent.php [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+use Cake\Http\Client;
+use InvalidArgumentException;
+use Psr\Http\Message\RequestInterface;
+
+/**
+ * Class Client Event
+ *
+ * @extends \Cake\Event\Event<\Cake\Http\Client>
+ */
+class ClientEvent extends Event
```
Comment (markstory) [inline src/Http/Client/ClientEvent.php:29]: This is our first foray into custom events. If this goes well we could consider it in other parts of the framework. I could see it helping provide better type/runtime safety with event payloads.

File: src/Http/Client.php (hunk 1 above)
Comment (ADmad) [reply src/Http/Client.php]: I too was leaning towards always calling `afterSend` but then kept it similar to model events where the "after" event is skipped if the result is returned by "before".

But as you said always calling afterSend makes more sense here, I'll update it.

File: src/Http/Client/ClientEvent.php (hunk 3 above)
Comment (ADmad) [reply src/Http/Client/ClientEvent.php:29]: I have always wanted to move towards using event specific classes but changing existing events would be significant amount of work so never took it on.

This feature request finally gave an opportunity to try out using a custom event class.

File: src/Http/Client.php (hunk 1 above)
Comment (othercorey) [reply src/Http/Client.php]: Maybe there can be a flag indicating that.

Comment (ADmad) [reply src/Http/Client.php]: You mean add an additional flag to the event when the response was returned by `beforeSend` instead of the actual call?

Comment (othercorey) [reply src/Http/Client.php]: Yes. if something like telemetry is important or tracking clean up post-call, probably helpful to know if the call was skipped.

**Review by markstory — APPROVED** [review]
Looks good to me.

**Pull request merged**